### DISCUSSION TITLE
JSS paper and possible library documentation

### DISCUSSION THREAD

**Discussion by ilyaZar** [rcppsmc/rcppsmc#73] [Ideas]
Hi all,

I am thinking about how to best approach the `JSS` paper writing, as @adamjohansen pointed out that we should focus on that getting done and implementing essential code along the line.

Since I am not part of the initial group that started the paper back then in 2017, I am very reluctant to simply add text or change work of others without explicit permission. My question would be on how are we best organizing this i.e. how you want me to contribute more specifically. For example

1. I could e.g. start by minor PRs to the private repo and simply write tiny bits that would change the paper slowly. 

2. Another way would be to add quite a lot of new text, and possibly chapter, and you could review the results ex-post, the changes being generally larger and more adhoc.

It probably depends on how much time we all want to invest in this, how much time we have to review new additions, and how much freedom you'd give me on changes or additions in general.

As for me, I have quite a lot of time at the disposal at the moment and would not mind to invest 1, maybe 2, months or so in writing polishing, and adding code when necessary. 

I am just not sure how to approach this best and any suggestions would help me a lot!

Have a nice weekend!

**Comment by adamjohansen** [comment]
I don't have any particularly strong view here.

I tend to think that if we're writing a paper together we need to trust one another not to do anything too crazy. If there are things anyone isn't sure about the best approach to, or if anything's not clear we can discuss or leave comments in the manuscript. I've not tended to worry about PRs with papers in the past and wouldn't personally have any objection to you pushing directly to the main branch (we can sort out permissions if needed) -- but I'm happy to do whatever works for everyone here.

I'm happy in principle to spend a bit of time on this although it's a limited resource so I might be a bit slow; it seems a better use of it than working on a code-base that isn't very widely used largely, I suspect, because of a lack of documentation (given that its predecessor still seems to get a bit of use) and I certainly wouldn't want you to feel that you have to do all of the writing here, but if you do have time and inspiration on your hands then it would be very welcome.

Incidentally, it's been a while since we did much with this and there might be a bit of maintenance to do to get it to build properly. I've added a missing "documentclass: jss" to the YAML frontmatter (needed with the newer version of rticles) but I'm still seeing some compile issues locally.

**Reply by LeahPrice** [reply]
Thanks @ilyaZar. It's fantastic that you're keen to work on this. I'm happy for you to push changes directly at will if everyone else is also.

Unfortunately I never pushed a huge chunk of changes I made back in 2018 in my fork, as I mentioned to you the other day. I'm going to do that and try to fix the compile issues while I'm at it now. Apologies everyone for the huge changes that will appear. Fortunately the version Ilya has been working on is the version from my repository with the extra changes (if I'm not misinterpreting).

**Reply by ilyaZar** [reply]
Hi @LeahPrice . Yes exactly, the version you sent me back the was exactly the one I was working/drafting some additional content for (which hopefully will be useful for the parer!).

**Comment by ilyaZar** [comment]
Thanks @adamjohansen !

In terms of permissions: so at least for me, it would help a lot, because having write access to `rcppsmc/post-gsoc` (or getting the OAuthentication request approved that I mentioned in an E-mail some time ago) allows to set up the overleaf-GitHub workflow I usually use to do all the academic paper writing projects.

**Comment by ilyaZar** [comment]
Thanks @LeahPrice for merging the PR, I just pushed part of my work as new PR but managed to include only additions to section 1.
I'll add to the other sections asap, but it is a bit time consuming since I have only hand-written notes on the working paper version.

If there is interest, in the last weeks I also worked on a general reference manual/API documentation in the "read-the-docs-style", that could complement the JSS paper work:

https://ilyazar.github.io/RcppSMCdocs/index.html

It is really easy to switch to classical doxygen style (if that is what you prefer @adamjohansen , since, I think the SMCTC library was in doxygen style). 

The workflow to update the documentation is almost automatic and described in the README of the project https://github.com/ilyaZar/RcppSMCdocs . Essentially, one updates the header files and cpp files of `RcppSMCdocs/include` and `RcppSMCdocs/src` of that repo (that currently contain plain copies of the files from `rcppsmc/inst/include` and `rcppsmc/src`) and then `git add/commit/push` triggers Github actions to update the github page from above.

_**The only non-automatic thing**_ is the current copy paste from `rcppsmc/inst/include` and `rcppsmc/src` to the corresponding folders `ilyaZar/RcppSMCdocs/inst` and `ilyaZar/RcppSMCdocs/src` in above repo. Is there a way to also automate the copying of files across Github repositories or something similar @eddelbuettel ? (copy pasting is certainly not a big thing, but one might forget a file so automation would be great!).

What do you think?

**Reply by eddelbuettel** [reply]
I may not have time dive in but if we already execute code in Actions we should also be able to copy files / deploy, no?

**Reply by ilyaZar** [reply]
I have some time and got your idea, thanks! (will have a look at `rcppsmc` yaml file and try to intertwine the commands there with mine)

**Reply by eddelbuettel** [reply]
I am not the most active Actions user (some people seem to go overboard) but I have test repos that commit too.  From my somewhat casual glances there seem to be Actions for just about anything.  Debugging can be a pain so by all means feel free to give up on anything "too crazy".

**Reply by adamjohansen** [reply]
Thanks, @ilyaZar this seems to fill a gap that has existed for _years_ and (admittedly superficially) looks great. I used doxygen because it seemed the cleanest solution to documenting code in 2005 (!) and I'm very happy to go along with whatever seems most convenient/user friendly in 2022 for this project.
If we can automate building the docs that would seem like a nice thing to do, but as @eddelbuettel suggests if it turns out not to be feasible to do in a sensible amount of time don't feel obliged.
Should we think about moving this repo to the RcppSMC org? (Just because it seems a more natural place for it to live.)

I would vote for removing the final paragraph on the index page:

> If you wish to be added to a mailing list to receive information about updates to this software, then please send an email to the author, [a.m.johansen@warwick.ac.uk](mailto:a.m.johansen%40warwick.ac.uk). Otherwise, please check this page regularly for updates.

because that mailing list doesn't really have any subscribed to it or get a lot of use these days. If we do keep it, we should rephrase it to avoid suggesting I'm responsible for the software... these days I'm sure others have contributed more code than me.

**Reply by ilyaZar** [reply]
@adamjohansen yes, I just removed what you asked me to remove and transferred the ownership to `rcppsmc`. I agree that this the the natural place it belongs to!

Probably it is also easier to align the Github actions from rcppsmc and those contained in that repo for copy/pasting the headers and cpp files we want to document.

**Reply by ilyaZar** [reply]
The new documentation page is now 

https://rcppsmc.github.io/RcppSMCdocs/

and we may want to change this in the repo description, top right corner in https://github.com/rcppsmc/RcppSMCdocs  (I forgot to change it before I transferred the ownership and now I can't anymore :) ).

**Reply by eddelbuettel** [reply]
Sorry.  We had goofed that up in the past and never cleaned up.  I just added your to RcppSMC.   You do have write right in RcppSMCdocs too so can you help me understand what you _cannot_ alter where?

**Reply by ilyaZar** [reply]
@eddelbuettel thanks! 

I can not change the link to the documentation page:

![image](https://user-images.githubusercontent.com/36898027/172912990-84f36005-17a8-40e0-aa92-e665a0aa414a.png)

it still points to https://ilyazar.github.io/RcppSMCdocs but should point to

https://rcppsmc.github.io/RcppSMCdocs

**Reply by LeahPrice** [reply]
Wow, thanks @ilyaZar! This looks great. I noticed that the copyright has my name first - I think that should be yours now :). I don't know anything about Github actions but it sounds like you've already got some good advice there.

**Comment by eddelbuettel** [comment]
What is the source URL / repo?  If if it (as I assume but you didn't say) https://github.com/rcppsmc/rcppsmc the all four of us have equal rights from what I see:

![image](https://user-images.githubusercontent.com/673121/172917229-4b233bf5-c8ef-4629-970e-77b2e734e844.png)

And if it is RcppSMCdocs then you seem to be the only one wit rights so colour me confused:

![image](https://user-images.githubusercontent.com/673121/172917615-a0e0b6a7-130e-4b28-8111-6b27f3b6338d.png)

On the other hand I was allowed to edit the link so that;s what I just did.  Confusing.

**Reply by adamjohansen** [reply]
I see the same so I'm equally confused.

**Reply by LeahPrice** [reply]
It looks like I can't access https://github.com/rcppsmc/rcppsmc/settings/access or any of the settings pages. I wonder if that's because Ilya and I are listed as members rather than owners according to the screenshot below. It's not causing me any issues at the moment but it might explain why Ilya doesn't have certain rights.
![image](https://user-images.githubusercontent.com/26055385/172997802-ebd252e9-9657-4f1e-9e0f-8b9150f8016c.png)

**Reply by ilyaZar** [reply]
@eddelbuettel thanks for changing the link and write permission! I can confirm that I can push/pull to RcppSMCdocs and the CI is triggered updating the docs as before! I'll also check the other repositories when we next update them!

@LeahPrice I think you are right and settings is for owners who can change repo-related things as the description or the link Dirk changed.

Write access probably relates to push/pull configurations and whether we can push to branches or have to do PRs. My guess is the reason why @eddelbuettel does not see anyone but me in the write access of `RcppSMCdocs` is probably that I am so far the only one who pushed there, but if anyone else of us would push there than that person is probably displayed as well. It is confusing though, I agree...

For me, it seems that everything is fixed now big thanks!
